File: LR8450/LR8450/LR8450_chinese/LR8450/chinese/Files/04/src/CommonSample/Visual Basic 2010/RS-232C_USB_Sample/Sample/Form1.vb
'*******************************************************************************
'このプログラムは、計測器に接続してコマンドの送受信を行います。
'コマンドの欄に送信したいコマンドを入力し、[送受信]ボタンを押すと送信されます。
'応答があるコマンド（?が含まれるコマンド）の場合は、テキストボックスに応答が表示されます。
'
'動作確認環境：Microsoft Visual Studio 2010
'              Version 10.0.40219.1 SP1Rel
'              Microsoft .NET Framework
'              Version 4.0.30319 SP1Rel
'              Microsoft Visual Basic 2010
'*******************************************************************************

Imports System.Diagnostics
Imports System.Text

Public Class Form1
    '---------------------------------------------------------------------------------------
    'フォームが開かれたときの処理
    '---------------------------------------------------------------------------------------
    Private Sub Form1_Load(sender As Object, e As System.EventArgs) Handles Me.Load
        'ボタンとテキストボックスの有効/無効の処理
        Button1.Enabled = True
        Button2.Enabled = False
        Button3.Enabled = False
        TextBox1.Enabled = True
        TextBox2.Enabled = True
        TextBox3.Enabled = False
        TextBox4.ReadOnly = True
        TextBox5.Enabled = True

    End Sub

    '---------------------------------------------------------------------------------------
    'ボタンを押したときの処理
    '---------------------------------------------------------------------------------------

    '「接続」ボタンを押したときの処理
    Private Sub Button1_Click(sender As System.Object, e As System.EventArgs) Handles Button1.Click
        '接続
        If OpenInterface(TextBox1.Text, TextBox2.Text, TextBox5.Text) = False Then
            Exit Sub
        End If

        'ボタンとテキストボックスの有効/無効の処理
        Button1.Enabled = False
        Button2.Enabled = True
        Button3.Enabled = True
        TextBox1.Enabled = False
        TextBox2.Enabled = False
        TextBox3.Enabled = True
        TextBox5.Enabled = False
    End Sub

    '「切断」ボタンを押したときの処理
    Private Sub Button2_Click(sender As System.Object, e As System.EventArgs) Handles Button2.Click
        '切断
        CloseInterface()

        'ボタンとテキストボックスの有効/無効の処理
        Button1.Enabled = True
        Button2.Enabled = False
        Button3.Enabled = False
        TextBox1.Enabled = True
        TextBox2.Enabled = True
        TextBox3.Enabled = False
        TextBox5.Enabled = True
    End Sub

    '「送受信」ボタンを押したときの処理
    Private Sub Button3_Click(sender As System.Object, e As System.EventArgs) Handles Button3.Click
        Button3.Enabled = False

        TextBox4.AppendText("<< " + TextBox3.Text + vbCrLf)                             'ログ出力
        If InStr(TextBox3.Text, "?") = 0 Then                                           'コマンドに?が含まれない場合は、コマンド送信のみ
            SendMsg(TextBox3.Text)                                                      'コマンド送信
        Else                                                                            'コマンドに?が含まれる場合は、コマンド送信と応答受信
            SendQueryMsg(TextBox3.Text, ReceiveTimeout)                                 'コマンド送信と応答受信
            TextBox4.AppendText(">> " + MsgBuf + vbCrLf)                                'ログ出力
        End If

        Button3.Enabled = True
    End Sub

    '「クリア」ボタンを押したときの処理
    Private Sub Button4_Click(sender As System.Object, e As System.EventArgs) Handles Button4.Click
        'テキストボックスの消去
        TextBox4.Clear()
    End Sub

    '---------------------------------------------------------------------------------------
    '通信インタフェース固有の処理
    '---------------------------------------------------------------------------------------

    '(0)クラス内変数
    Private SerialPort As System.IO.Ports.SerialPort                                    'シリアルポートデバイス
    Private MsgBuf As String = ""                                                       '受信データ
    Private ReceiveTimeout As Long = 0                                                  '受信タイムアウト時間（ms）

    '(1)接続
    Private Function OpenInterface(ByVal port As String, ByVal speed As String, ByVal timeout As String) As Boolean
        Dim ret As Boolean = False

        Try
            ReceiveTimeout = CInt(timeout) * 1000
            SerialPort = New System.IO.Ports.SerialPort                                 'シリアルポートオブジェクトを作成
            SerialPort.PortName = port                                                  'COMポートを設定
            SerialPort.BaudRate = Convert.ToInt32(speed)                                'スピードを設定
            SerialPort.Open()                                                           'シリアルポートオープン
            ret = True
        Catch Ex As Exception
            MsgBox(Ex.Message)
        End Try

        OpenInterface = ret
    End Function

    '(2)切断
    Private Function CloseInterface() As Boolean
        Dim ret As Boolean = False

        Try
            If SerialPort.IsOpen Then
                SerialPort.Close()                                                      'シリアルポートクローズ
            End If
            SerialPort.Dispose()                                                        'シリアルポートオブジェクトを廃棄
            ret = True
        Catch ex As Exception
            MsgBox(ex.Message)
        End Try

        CloseInterface = ret
    End Function

    '(3)コマンド送信
    Private Function SendMsg(ByVal strMsg As String) As Boolean
        Dim ret As Boolean = False

        Try
            strMsg += vbCrLf                                                            'ターミネータ「CR+LF」を付加
            SerialPort.Write(strMsg)                                                    '送信バッファに書き込み
            ret = True
        Catch Ex As Exception
            MsgBox(Ex.Message)
        End Try

        SendMsg = ret
    End Function

    '(4)受信
    Private Function ReceiveMsg(ByVal timeout As Long) As Boolean
        Dim ret As Boolean = False
        Dim rcv As String = ""
        Dim buf As New StringBuilder()
        Dim sw As New Stopwatch()

        Try
            MsgBuf = ""                                                                 '受信データをクリア

            sw.Start()                                                                  'タイムアウト用ストップウォッチを開始
            'ターミネータ「LF」を受信するまでループ
            Do
                If SerialPort.BytesToRead > 0 Then                                      '受信バッファにデータがあれば読み取り
                    rcv = SerialPort.ReadExisting()                                     '受信バッファから読み取り
                    rcv = rcv.Replace(vbCr, "")                                         '受信データ内の「CR」を削除
                    If rcv.IndexOf(vbLf) >= 0 Then                                      'ターミネータ「LF」を受信したら終了
                        rcv = rcv.Substring(0, rcv.IndexOf(vbLf))                       '受信データを「LF」の手前までで切り詰め
                        buf.Append(rcv)                                                 '受信データを保存
                        MsgBuf = buf.ToString()
                        Exit Do
                    Else
                        buf.Append(rcv)                                                 '受信データを保存
                    End If
                End If
                'タイムアウト処理
                If sw.ElapsedMilliseconds > timeout Then
                    MsgBuf = "Timeout"
                    MsgBox(MsgBuf)
                    ReceiveMsg = ret
                    Exit Function
                End If
            Loop
            sw.Stop()                                                                   'ストップウォッチを停止
            ret = True

        Catch Ex As Exception
            MsgBuf = "Error"
            MsgBox(Ex.Message)
        End Try

        ReceiveMsg = ret
    End Function

    '(5)コマンド送受信
    Private Function SendQueryMsg(ByVal strMsg As String, ByVal timeout As Long) As Boolean
        Dim ret As Boolean = False

        ret = SendMsg(strMsg)                                                           'コマンド送信
        If ret Then
            ret = ReceiveMsg(timeout)                                                   '送信が成功したら応答を受信
        End If

        SendQueryMsg = ret
    End Function

End Class

File: LR8450/LR8450/LR8450_chinese/LR8450/english/Files/04/src/CommonSample/Visual C# 2010/RS-232C_USB_Sample/Sample/Form1.cs
//*******************************************************************************
//このプログラムは、計測器に接続してコマンドの送受信を行います。
//コマンドの欄に送信したいコマンドを入力し、[送受信]ボタンを押すと送信されます。
//応答があるコマンド（?が含まれるコマンド）の場合は、テキストボックスに応答が表示されます。
//
//動作確認環境：Microsoft Visual Studio 2010
//				Version 10.0.40219.1 SP1Rel
//				Microsoft .NET Framework
//				Version 4.0.30319 SP1Rel
//				Microsoft Visual C# 2010
//*******************************************************************************

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace Sample
{
	public partial class Form1 : Form
	{
		//---------------------------------------------------------------------------------------
		//フォームが開かれたときの処理
		//---------------------------------------------------------------------------------------
		public Form1()
		{
			InitializeComponent();
			//ボタンとテキストボックスの有効/無効の処理
			Button1.Enabled = true;
			Button2.Enabled = false;
			Button3.Enabled = false;
			TextBox1.Enabled = true;
			TextBox2.Enabled = true;
			TextBox3.Enabled = false;
			TextBox4.ReadOnly = true;
			TextBox5.Enabled = true;
		}

		//---------------------------------------------------------------------------------------
		//ボタンを押したときの処理
		//---------------------------------------------------------------------------------------

		//「接続」ボタンを押したときの処理
		private void Button1_Click(object sender, EventArgs e)
		{
			//接続
			if (OpenInterface(TextBox1.Text, TextBox2.Text, TextBox5.Text) == false)
			{
				return;
			}

			//ボタンとテキストボックスの有効/無効の処理
			Button1.Enabled = false;
			Button2.Enabled = true;
			Button3.Enabled = true;
			TextBox1.Enabled = false;
			TextBox2.Enabled = false;
			TextBox3.Enabled = true;
			TextBox5.Enabled = false;
		}

		//「切断」ボタンを押したときの処理
		private void Button2_Click(object sender, EventArgs e)
		{
			//切断
			CloseInterface();

			//ボタンとテキストボックスの有効/無効の処理
			Button1.Enabled = true;
			Button2.Enabled = false;
			Button3.Enabled = false;
			TextBox1.Enabled = true;
			TextBox2.Enabled = true;
			TextBox3.Enabled = false;
			TextBox5.Enabled = true;
		}

		//「送受信」ボタンを押したときの処理
		private void Button3_Click(object sender, EventArgs e)
		{
			Button3.Enabled = false;

			TextBox4.AppendText("<< " + TextBox3.Text + "\r\n");						//ログ出力
			if (TextBox3.Text.Contains("?") == false)									//コマンドに?が含まれない場合は、コマンド送信のみ
			{				
				SendMsg(TextBox3.Text); 												//コマンド送信
			}																			//コマンドに?が含まれる場合は、コマンド送信と応答受信
			else
			{
				SendQueryMsg(TextBox3.Text, ReceiveTimeout);							//コマンド送信と応答受信
				TextBox4.AppendText(">> " + MsgBuf + "\r\n");							//ログ出力
			}

			Button3.Enabled = true;
		}

		//「クリア」ボタンを押したときの処理
		private void Button4_Click(object sender, EventArgs e)
		{
			//テキストボックスの消去
			TextBox4.Clear();
		}

		//---------------------------------------------------------------------------------------
		//通信インタフェース固有の処理
		//---------------------------------------------------------------------------------------

		//(0)クラス内変数
		private System.IO.Ports.SerialPort SerialPort;									//シリアルポートデバイス
		private string MsgBuf = ""; 													//受信データ
		private long ReceiveTimeout = 0;												//受信タイムアウト時間（ms）

		//(1)接続
		private bool OpenInterface(string port, string speed, string timeout)
		{
			bool ret = false;

 			try
			{
				ReceiveTimeout = Convert.ToInt64(timeout) * 1000;
				SerialPort = new System.IO.Ports.SerialPort();							//シリアルポートオブジェクトを作成
				SerialPort.PortName = port; 											//COMポートを設定
				SerialPort.BaudRate = Convert.ToInt32(speed);							//スピードを設定
				SerialPort.Open();														//シリアルポートオープン
				ret = true;
			}
			catch(Exception e)
			{
				MessageBox.Show(e.Message);
			}

			return ret;
		}

		//(2)切断
		private bool CloseInterface()
		{
			bool ret = false;

			try
			{
				if (SerialPort.IsOpen)
				{
					SerialPort.Close();													//シリアルポートクローズ
				}
				SerialPort.Dispose();													//シリアルポートオブジェクトを廃棄
				ret = true;
			}
			catch(Exception e)
			{
				MessageBox.Show(e.Message);
			}

			return ret;
		}

		//(3)コマンド送信
		private bool SendMsg(string strMsg)
		{
			bool ret = false;

			try
			{
				strMsg += "\r\n";														//ターミネータ「CR+LF」を付加
				SerialPort.Write(strMsg);												//送信バッファに書き込み
				ret = true;
			}
			catch(Exception e)
			{
				MessageBox.Show(e.Message);
			}

			return ret;
		}

		//(4)受信
		private bool ReceiveMsg(long timeout)
		{
			bool ret = false;
			string rcv = "";
			StringBuilder buf = new StringBuilder();
			Stopwatch sw = new Stopwatch();

			try
			{
				MsgBuf = "";															//受信データをクリア

				sw.Start();																//タイムアウト用ストップウォッチを開始
				//ターミネータ「LF」を受信するまでループ
				while(true){
					if (SerialPort.BytesToRead > 0)										//受信バッファにデータがあれば読み取り
					{
						rcv = SerialPort.ReadExisting();								//受信バッファから読み取り
						rcv = rcv.Replace("\r", "");									//受信データ内の「CR」を削除
						if (rcv.IndexOf("\n") >= 0)										//ターミネータ「LF」を受信したら終了
						{
							rcv = rcv.Substring(0, rcv.IndexOf("\n"));					//受信データを「LF」の手前までで切り詰め
							buf.Append(rcv);											//受信データを保存
							MsgBuf = buf.ToString();
							break;
						}
						else
						{
							buf.Append(rcv);											//受信データを保存
						}
					}
					//タイムアウト処理
					if (sw.ElapsedMilliseconds > timeout)
					{
						MsgBuf = "Timeout";
						MessageBox.Show(MsgBuf);
						return ret;
					}
				}
				sw.Stop();																//ストップウォッチを停止
				ret = true;
			}
			catch(Exception e)
			{
				MsgBuf = "Error";
				MessageBox.Show(e.Message);
			}

			return ret;
		}

		//(5)コマンド送受信
		private bool SendQueryMsg(string strMsg, long timeout)
		{
			bool ret = false;

			ret = SendMsg(strMsg);														//コマンド送信
			if (ret)
			{
				ret = ReceiveMsg(timeout);												//送信が成功したら応答を受信
			}

			return ret;
		}

	}
}
